src/plugins/intel_cpu/src/utils/serialize.cpp

```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "serialize.hpp"
+
+#include "openvino/core/descriptor_tensor.hpp"
+#include "openvino/core/parallel.hpp"
+#include "openvino/runtime/shared_buffer.hpp"
+#include "utils/codec_xor.hpp"
+
+namespace ov {
+namespace intel_cpu {
+
+////////// ModelSerializer //////////
+
+ModelSerializer::ModelSerializer(std::ostream& ostream, CacheEncrypt encrypt_fn)
+    : m_ostream(ostream), m_cache_encrypt(std::move(encrypt_fn)) {}
+
+void ModelSerializer::operator<<(const std::shared_ptr<ov::Model>& model) {
+    auto serialize_info = [&](std::ostream& stream) {
+        pugi::xml_document xml_doc;
+        pugi::xml_node root = xml_doc.append_child("cnndata");
+        pugi::xml_node outputs = root.append_child("outputs");
+        for (const auto& out : model->get_results()) {
+            auto out_node = outputs.append_child("out");
+            const auto name = ov::descriptor::get_ov_tensor_legacy_name(out->input_value(0).get_tensor());
+            out_node.append_attribute("name").set_value(name.c_str());
+        }
+        xml_doc.save(stream);
+    };
+
+    ov::pass::StreamSerialize serializer(m_ostream, serialize_info, m_cache_encrypt);
+    serializer.run_on_model(std::const_pointer_cast<ov::Model>(model->clone()));
+}
+
+////////// ModelDeserializer //////////
+
+ModelDeserializer::ModelDeserializer(std::istream& model_stream, ModelBuilder fn, CacheDecrypt decrypt_fn)
+    : m_istream(model_stream), m_model_builder(std::move(fn)), m_cache_decrypt(std::move(decrypt_fn)) {}
+
+void ModelDeserializer::set_info(pugi::xml_node& root, std::shared_ptr<ov::Model>& model) {
+    pugi::xml_node outputs = root.child("outputs");
+    auto nodes_it = outputs.children("out").begin();
+    size_t size = model->outputs().size();
+    for (size_t i = 0lu; i < size; ++nodes_it, i++) {
+        std::string name = nodes_it->attribute("name").value();
+        if (name.empty())
+            continue;
+        auto result = model->output(i).get_node_shared_ptr();
+        ov::descriptor::set_ov_tensor_legacy_name(result->input_value(0).get_tensor(), name);
+    }
+}
+
+void ModelDeserializer::operator>>(std::shared_ptr<ov::Model>& model) {
+    if (auto mmap_stream = dynamic_cast<MmapStream*>(&m_istream)) {
+        process_mmap(model, mmap_stream->m_memory);
+    } else {
+        process_stream(model);
+    }
+}
```

Comment:
This is really bad approach, as it brings implicit coupling between the plugin deserializer, frontend and core levels. We need really strong justification to allow such design flaws to be introduced.
A better solution would be an extension of the necessary interfaces with explicit passing FS mapped memory.